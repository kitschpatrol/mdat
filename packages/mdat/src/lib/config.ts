/* eslint-disable complexity */

import { mdatJsonLoader } from './mdat-json-loader'
import { findPackage } from './utilities'
import chalk from 'chalk'
import { type CosmiconfigResult, cosmiconfig } from 'cosmiconfig'
import fs from 'node:fs/promises'
import path from 'node:path'
import plur from 'plur'
import { type NormalizedPackageJson, readPackage } from 'read-pkg'
import {
	type Options,
	type Rules,
	deepMergeDefined,
	log,
	optionsSchema,
	rulesSchema,
} from 'remark-mdat'
import { type JsonValue } from 'type-fest'
import { z } from 'zod'

export type Config = {
	assetsPath?: string
	packageFile?: string
} & Options

// Reflect defaults and normalization
export type ConfigLoaded = {
	addMetaComment: boolean
	assetsPath: string
	closingPrefix: string
	keywordPrefix: string
	metaCommentIdentifier: string
	packageFile: string | undefined
	rules: Rules
}

const configSchema = optionsSchema
	.merge(
		z.object({
			assetsPath: z.string().optional(),
			packageFile: z.string().optional(),
		}),
	)
	.describe('Config Extension')

/**
 * Generously accept either string paths to .ts, .js, or .json files with
 * `Config` type default exports. Takes a single value, or an array of values which
 * will be merged right to left.
 */
export type ConfigToLoad = Array<Config | string> | Config | string

/**
 * Generously accept either string paths to .ts, .js, or .json files with
 * `Rules` type default exports.
 */
export type RulesToLoad = Array<Rules | string> | Rules | string

// Nomenclature is somewhat obtuse:
// Config:
// File loaded implicitly by cosmiconfig, or passed in as an object or string path to load. Has top-level settings options, plus an optional rules object of its own.
//
// Rules:
// File loaded explicitly by the user, it's a record where each key is a replacement keyword
//
// Rules are merged into a final internal config file

/**
 * Load and validate mdat configuration / rule sets
 * Uses cosmiconfig to search in the usual places.
 * Merge precedence: Additional Config Paths < Search Path < mdat-remark defaults
 *
 * Generic to accommodate additional Config options, so set T to your custom config type if needed. You must provide a matching configExtensionSchema as well.
 */
export async function loadConfig(options?: {
	/**
	 * Additional Config objects to merge.
	 *
	 * Strings are treated as paths to `ts`, `js`, or `json` files with `Config` type default exports. These will be dynamically loaded by Cosmiconfig.
	 * Accepts an individual item, or an array. Objects in the array will be merged right to left.
	 *
	 */
	additionalConfig?: ConfigToLoad
	/**
	 * Additional Rules objects to merge.
	 *
	 * Strings are treated as paths to `ts`, `js`, or `json` files with `Rules` type default exports. These will be dynamically loaded by Cosmiconfig.
	 * Accepts an individual item, or an array. Objects in the array will be merged right to left, and take precedence over any rules in previously loaded Config objects as well.
	 */
	additionalRules?: RulesToLoad
	/** Search for config in specific directories, mainly useful for testing. Cosmiconfig default search paths used if unset. */
	searchFrom?: string
}): Promise<ConfigLoaded> {
	const { additionalConfig, additionalRules, searchFrom } = options ?? {}
	// Set defaults, remark-mdat sets these as well... but it sets them at runtime,
	// we need to set them here so they're on the returned object
	let finalConfig: Config = {
		addMetaComment: false,
		assetsPath: './assets',
		closingPrefix: '/',
		keywordPrefix: '',
		metaCommentIdentifier: '+',
		packageFile: await findPackage(),
		rules: {
			mdat: `Powered by the Markdown Autophagic Template system: [mdat](https://github.com/kitschpatrol/mdat).`,
		},
	}

	// 1. Search and load cosmiconfig locations
	const configExplorer = cosmiconfig('mdat')
	const results = await configExplorer.search(searchFrom)

	if (results) {
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		const { config, filepath } = results
		log.info(`Using config from "${filepath}"`)
		const configFromObject = getAndValidateConfigFromObject(config, configSchema)
		if (configFromObject) {
			finalConfig = deepMergeDefined(finalConfig, configFromObject)
		}
	}

	// 3. Load and merge additional configs
	if (additionalConfig !== undefined) {
		// Ensure we have an array to work with
		const additionalConfigsArray = Array.isArray(additionalConfig)
			? additionalConfig
			: [additionalConfig]

		for (const configOrPath of additionalConfigsArray) {
			let config: unknown

			if (typeof configOrPath === 'string') {
				// It's probably a path to a config object
				const results = await configExplorer.load(configOrPath)
				if (results === null || results === undefined) continue
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
				const { config: loadedConfig, filepath } = results
				log.info(`Loaded additional config from "${filepath}"`)
				config = loadedConfig
			} else {
				// It's probably a pre-loaded config object
				config = configOrPath
			}

			if (config === undefined) continue

			log.info('Merging configuration object')
			const configFromObject = getAndValidateConfigFromObject(config, configSchema)

			if (configFromObject !== undefined) {
				finalConfig = deepMergeDefined(finalConfig, configFromObject)
			}
		}
	}

	// 4. Load and merge additional rule sets
	if (additionalRules !== undefined) {
		// Ensure we have an array to work with
		const additionalRulesArray = Array.isArray(additionalRules)
			? additionalRules
			: [additionalRules]

		// Create new cosmiconfig loader with custom handling
		// to flatten .json files into rule sets
		const rulesExplorer = cosmiconfig('mdat', {
			loaders: {
				'.json': mdatJsonLoader,
			},
		})

		for (const rulesOrPath of additionalRulesArray) {
			let rules: unknown

			if (typeof rulesOrPath === 'string') {
				// It's a path to a rules object (we hope)
				let results: CosmiconfigResult
				// Special work-around for Cosmiconfig's rather zealous package.json interception
				// So we can treat a package.json as a rule set... but note that this
				// is different from how additionalConfigs is handles, where a _configuration_ object
				// is loaded from package.json, instead of being interpreted as _rules_.
				if (path.basename(rulesOrPath).endsWith('package.json')) {
					const packageJson = await fs.readFile(rulesOrPath, 'utf8')
					const flatJson = mdatJsonLoader(rulesOrPath, packageJson) as JsonValue
					results = {
						config: flatJson,
						filepath: rulesOrPath,
					}
				} else {
					// Let cosmiconfig call the custom loader itself
					results = await rulesExplorer.load(rulesOrPath)
				}

				if (results === null || results === undefined) continue
				// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
				const { config: loadedRules, filepath } = results
				log.info(`Loaded additional config from "${filepath}"`)
				rules = loadedRules
			} else {
				// It's a pre-loaded rules object
				rules = rulesOrPath
			}

			if (rules === undefined) continue

			log.info('Merging rules into configuration object')
			const configFromRulesObject = getAndValidateConfigFromRulesObject(rules, rulesSchema)

			if (configFromRulesObject !== undefined) {
				finalConfig = deepMergeDefined(finalConfig, configFromRulesObject)
			}
		}
	}

	if (finalConfig.rules) {
		const prettyRules = Object.keys(finalConfig.rules)
			.sort()
			.map((rule) => `"${chalk.bold.green(rule)}"`)
		log.info(
			`Loaded ${chalk.bold(prettyRules.length)} mdat comment expansion ${plur('rule', prettyRules.length)}:`,
		)
		for (const rule of prettyRules) {
			log.info(`\t${rule}`)
		}
	} else {
		log.error('No rules loaded from additional configurations or rules, using default.')
	}

	// Trust that all the non-optional values have been set

	// Set global for rules
	config = finalConfig as ConfigLoaded

	return finalConfig as ConfigLoaded
}

// Helpers

function getAndValidateConfigFromRulesObject(
	rulesObject: unknown,
	rulesSchema: z.ZodSchema,
): Config | undefined {
	if (rulesSchema.safeParse(rulesObject).success) {
		// Wrap the rules object in a config object
		const config = {
			rules: rulesObject as Rules,
		}
		return config as Config
	}

	log.error(
		`Rules object has the wrong shape. Ignoring and using default configuration:\n${JSON.stringify(rulesObject, undefined, 2)}`,
	)
}

function getAndValidateConfigFromObject(
	configObject: unknown,
	configSchema: z.ZodSchema,
): Config | undefined {
	if (configSchema.safeParse(configObject).success) {
		return configObject as Config
	}

	log.error(
		`Config object has the wrong shape. Ignoring and using default configuration:\n${JSON.stringify(configObject, undefined, 2)}`,
	)
}

// Rule helpers

// Caution: Impure functions

// TODO better to pass config into rules?
// Let rules access config for custom behavior
let config: ConfigLoaded | undefined
export async function getConfig(): Promise<ConfigLoaded> {
	// Defaults guarantee all keys will be defined
	// Readme config should be defined at this point when called from one of the
	// API functions that handle config loading, but if not, e.g. in testing,
	// we can load it here

	if (config === undefined) {
		log.warn('getConfig(): config was undefined')
		config ??= await loadConfig()
	}

	return config
}

// Convenience function for rules
// Load as package json only as needed, memoize
// Rules could call this themselves, but this is more convenient and efficient
let packageJson: NormalizedPackageJson | undefined
export async function getPackageJson(): Promise<NormalizedPackageJson> {
	const { packageFile } = await getConfig()
	if (packageFile === undefined) {
		throw new Error('No packageFile found or set in config')
	}

	packageJson ??= await readPackage({ cwd: path.dirname(packageFile) })
	if (packageJson === undefined) {
		throw new Error('No package.json found')
	}

	return packageJson
}
