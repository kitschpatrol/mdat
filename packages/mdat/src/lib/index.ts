export {
	type ExpandFileOptions,
	type ExpandFilesOptions,
	type ExpandStringOptions,
	expandFile,
	expandFiles,
	expandString,
} from './api'
export { loadConfig } from './config'
export {
	type NormalizedRules,
	type Options,
	type Rules,
	getMdatReports,
	log,
	reporterMdat,
} from 'remark-mdat'
