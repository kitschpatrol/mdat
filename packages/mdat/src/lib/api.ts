import { type ConfigToLoad, type RulesToLoad, loadConfig } from './config'
import {
	getCheckProcessor,
	getCleanProcessor,
	getExpandProcessor,
	processFiles,
	processString,
} from './processors'
import { type VFile } from 'vfile'

/**
 * Writing is the responsibility of the caller (e.g. via `await write(result)`)
 */
export async function expandFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getExpandProcessor, name, output, config, rules)
}

export async function expandString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getExpandProcessor, config, rules)
}

export async function checkFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getCheckProcessor, name, output, config, rules)
}

export async function checkString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getCheckProcessor, config, rules)
}

export async function collapseFiles(
	files: string | string[],
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile[]> {
	return processFiles(files, loadConfig, getCleanProcessor, name, output, config, rules)
}

export async function collapseString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	return processString(markdown, loadConfig, getCleanProcessor, config, rules)
}
