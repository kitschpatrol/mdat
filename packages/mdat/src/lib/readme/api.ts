import { cleanFile, expandFile, expandString } from '../api'
import {
	type Config,
	type ConfigLoaded,
	type ConfigToLoad,
	type RulesToLoad,
	loadConfig,
} from '../config'
import readmeRules from './rules'
import { type Simplify } from 'type-fest'
import { type VFile } from 'vfile'

// Readme and package files are mandatory for readme operations
type ReadmeConfigLoaded = Simplify<
	{
		packageFile: string
		readmeFile: string
	} & ConfigLoaded
>

export async function expandReadme(
	name?: string,
	output?: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})

	return expandFile(resolvedConfig.readmeFile, name, output, resolvedConfig)
}

export async function expandReadmeString(
	markdown: string,
	config?: ConfigToLoad,
	rules?: RulesToLoad,
): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})

	return expandString(markdown, resolvedConfig)
}

export async function cleanReadme(config?: ConfigToLoad): Promise<VFile> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
	})

	return cleanFile(resolvedConfig.readmeFile, undefined, undefined, resolvedConfig)
}

// Helpers

// Convenience loader to always include the default readme config
type LoadConfigOptions = Parameters<typeof loadConfig>[0]
export async function loadConfigReadme(options?: LoadConfigOptions): Promise<ReadmeConfigLoaded> {
	const defaultReadmeConfig: Config = {
		addMetaComment: true,
		rules: readmeRules,
	}

	const { additionalConfig = [], ...rest } = options ?? {}
	const additionalConfigArray = Array.isArray(additionalConfig)
		? additionalConfig
		: [additionalConfig]

	const result = await loadConfig({
		additionalConfig: [defaultReadmeConfig, ...additionalConfigArray],
		...rest,
	})

	// This should never happen because the defaults are set in loadConfigReadme
	if (result.packageFile === undefined || result.readmeFile === undefined) {
		throw new Error('Package and readme files are required in `mdat readme` config')
	}

	// Trust all optionals have been set
	return result as ReadmeConfigLoaded
}
