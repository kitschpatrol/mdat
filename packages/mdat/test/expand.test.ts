import { expandString } from '../src/lib/api'
import fs from 'node:fs/promises'
import { describe, expect, it } from 'vitest'

describe('comment expansion', () => {
	it('should expand comments', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const expandedString = await expandString(markdown, {
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should be idempotent', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const firstPass = await expandString(markdown, {
			rules: ['./test/assets/test-rules.js'],
		})

		const secondPass = await expandString(firstPass, {
			rules: ['./test/assets/test-rules.js'],
		})

		expect(firstPass).toEqual(secondPass)
	})

	it('should expand prefixed comments only', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const expandedString = await expandString(markdown, {
			keywordPrefix: 'mm-',
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should include the meta tag if asked', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const expandedString = await expandString(markdown, {
			addMetaComment: true,
			rules: ['./test/assets/test-rules.js'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should throw an error if rule set is invalid', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')

		await expect(
			expandString(markdown, {
				addMetaComment: true,
				rules: ['./test/assets/test-rules-invalid.js'],
			}),
		).rejects.toThrow()
	})

	it('should work with hand-crafted json rules', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const expandedString = await expandString(markdown, {
			addMetaComment: true,
			rules: ['./test/assets/test-rules-json.json'],
		})

		expect(expandedString).toMatchSnapshot()
	})

	it('should work with arbitrary json files', async () => {
		const markdown = await fs.readFile('./test/assets/test-document.md', 'utf8')
		const expandedString = await expandString(markdown, {
			addMetaComment: true,
			rules: ['./package.json'],
		})

		expect(expandedString).toMatchSnapshot()
	})
})
