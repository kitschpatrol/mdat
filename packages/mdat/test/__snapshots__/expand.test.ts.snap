// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment expansion > should expand comments 1`] = `
"# The Mdat sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Basic comment clean up

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.js

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

ðŸŒ³ I am between two ferns ðŸŒ³

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, Mdat will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and Mdat will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

I had to be last

<!-- /basic-ordered -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->test-rules.js<!-- /basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

I am a paragraph 1

I am a paragraph 2

I am a paragraph 3

<!-- /basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->
>
> I am a paragraph 1
>
> I am a paragraph 2
>
> I am a paragraph 3
>
> <!-- /basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><b>Absolutely</b><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options -->

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Adjacent top-level expansions with leading space

 <!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Expansion with leading space

 <!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->
"
`;

exports[`comment expansion > should expand prefixed comments only 1`] = `
"# The Mdat sample document

## Basic comment expansion

<!-- basic -->

## Basic comment clean up

<!-- basic -->

Stale content that will be replaced

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

**A bold statement from test-rules.js**

<!-- /mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, Mdat will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and Mdat will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

<!-- basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic --><!-- basic -->Z

## Adjacent top-level expansions

<!-- basic --><!-- basic -->Z

## Adjacent top-level expansions

<!-- basic --><b>Absolutely</b><!-- basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Adjacent top-level expansions

<!-- basic --><!-- basic -->

## Adjacent top-level expansions with leading space

 <!-- basic --><!-- basic -->

## Expansion with leading space

 <!-- basic -->
"
`;

exports[`comment expansion > should include the meta tag if asked 1`] = `
"<!-- Warning: Content in HTML comment blocks generated by mdat on 2024-01-28. -->

# The Mdat sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Basic comment clean up

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

I was generated from test-rules.js

<!-- /basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

ðŸŒ³ I am between two ferns ðŸŒ³

<!-- /basic-options -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

ðŸª´ I am between two ferns ðŸª´

<!-- /basic-options -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, Mdat will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and Mdat will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

I had to be last

<!-- /basic-ordered -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->test-rules.js<!-- /basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

I am a paragraph 1

I am a paragraph 2

I am a paragraph 3

<!-- /basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->
>
> I am a paragraph 1
>
> I am a paragraph 2
>
> I am a paragraph 3
>
> <!-- /basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

<!-- basic -->

**A bold statement from test-rules.js**

<!-- /basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><b>Absolutely</b><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->ðŸª´ I am between two ferns ðŸª´<!-- /basic-options -->

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Adjacent top-level expansions with leading space

 <!-- basic -->**A bold statement from test-rules.js**<!-- /basic --><!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->

## Expansion with leading space

 <!-- basic -->**A bold statement from test-rules.js**<!-- /basic -->
"
`;

exports[`comment expansion > should work with arbitrary json files 1`] = `
"<!-- Warning: Content in HTML comment blocks generated by mdat on 2024-01-28. -->

# The Mdat sample document

## Basic comment expansion

<!-- basic -->

## Basic comment clean up

<!-- basic -->

Stale content that will be replaced

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

mdat

<!-- /name -->

<!--- author.name -->

Eric Mika

<!-- /author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, Mdat will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and Mdat will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

<!-- basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic --><!-- basic -->Z

## Adjacent top-level expansions

<!-- basic --><!-- basic -->Z

## Adjacent top-level expansions

<!-- basic --><b>Absolutely</b><!-- basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Adjacent top-level expansions

<!-- basic --><!-- basic -->

## Adjacent top-level expansions with leading space

 <!-- basic --><!-- basic -->

## Expansion with leading space

 <!-- basic -->
"
`;

exports[`comment expansion > should work with hand-crafted json rules 1`] = `
"<!-- Warning: Content in HTML comment blocks generated by mdat on 2024-01-28. -->

# The Mdat sample document

## Basic comment expansion

<!-- basic -->

**A bold statement from test-rules-json.json**

<!-- /basic -->

## Basic comment clean up

<!-- basic -->

**A bold statement from test-rules-json.json**

<!-- /basic -->

## Prefixed comment

Expansion only happens when \`--prefix mm\` option is set:

<!-- mm-basic -->

## Comment expansion with dynamic content

Shows the name of the rules file:

<!-- basic-dynamic -->

## Required expansion

Inclusion is validated via \`--check\` option

<!-- basic-list-required -->

* I
* am
* a
* list
* that
* must
* be
* here

<!-- /basic-list-required -->

## Expansions with options argument, canonical syntax

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Same expansion with different options

<!-- basic-options {prefix: "ðŸŒ³ ", suffix: " ðŸŒ³"} -->

## Expansions with arguments, syntax forgiveness

<!--- # basic-options({prefix: "ðŸª´ ", suffix: " ðŸª´"}) -->

## Expansions via dot paths given arbitrary json files as rules

Given e.g. \`--rules ./package.json\`, allow access to fields in the file:

<!--- name -->

<!--- author.name -->

## Handling empty rule content

This expansion rule always returns an empty string. The comment will not be expanded, Mdat will report the error.

<!-- basic-empty -->

## Handling non-required rule errors

This expansion rule always throws an error. The comment will not be expanded, and Mdat will report the error.

<!-- basic-throws -->

## Ordered expansion

Order in the document is validated via \`--check\` option

<!-- basic-ordered -->

## Inline expansions with paragraph elements

This is a sentence with a "*<!-- basic-inline -->*" inside of it.

## Multiple paragraphs

<!-- basic-multiple-paragraphs -->

## Multiple paragraphs blockquote

> <!-- basic-multiple-paragraphs -->

## Tags in code fence

Should not expand

\`\`\`html
<!-- basic -->
\`\`\`

## Stacked expansions

<!-- basic -->

**A bold statement from test-rules-json.json**

<!-- /basic -->

<!-- basic -->

**A bold statement from test-rules-json.json**

<!-- /basic -->

The following required an extra step to split html statements that were being interpreted as contiguous

## Adjacent inline expansions

A<!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic --><!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic --><!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->Z

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic --><b>Absolutely</b><!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->

## Adjacent top-level expansions with options

<!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} --><!-- basic-options {prefix: "ðŸª´ ", suffix: " ðŸª´"} -->

## Adjacent top-level expansions

<!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic --><!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->

## Adjacent top-level expansions with leading space

 <!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic --><!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->

## Expansion with leading space

 <!-- basic -->**A bold statement from test-rules-json.json**<!-- /basic -->
"
`;
